[SPARK-29691][ML][PYTHON] ensure Param objects are valid in fit, transform

JohnHBauer opened the pull request:
modify Param._copyValues to check valid Param objects supplied as extra

### What changes were proposed in this pull request?

Estimator.fit() and Model.transform() accept a dictionary of extra parameters whose values are used to overwrite those supplied at initialization or by default.  Additionally, the ParamGridBuilder.addGrid accepts a parameter and list of values. The keys are presumed to be valid Param objects. This change adds a check that only Param objects are supplied as keys.

### Why are the changes needed?

Param objects are created by and bound to an instance of Params (Estimator, Model, or Transformer). They may be obtained from their parent as attributes, or by name through getParam.

The documentation does not state that keys must be valid Param objects, nor describe how one may be obtained. The current behavior is to silently ignore keys which are not valid Param objects.

### Does this PR introduce any user-facing change?

If the user does not pass in a Param object as required for keys in `extra` for Estimator.fit() and Model.transform(), and `param` for ParamGridBuilder.addGrid, an error will be raised indicating it is an invalid object. 

### How was this patch tested?

Added method test_copy_param_extras_check to test_param.py.   Tested with Python 3.7

JohnHBauer committed:
ensure Param objects are valid in fit, transform

modify Param._copyValues to check valid Param objects supplied as extra

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

JohnHBauer committed:
Update PULL_REQUEST_[SPARK-29691]

JohnHBauer committed:
Update __init__.py

JohnHBauer committed:
Update __init__.py

huaxingao commented:
@JohnHBauer Thanks for the pull request!
Seems to me that @BryanCutler meant not to extend the API to accept parameter names, but just raise an Error if the values in extra are not an instance of Param. 
Also for type checking, seems to me that it might be more straightforward to do type checking and raise error in ```fit```/```transform``` instead of going down to ```_copyValues```

JohnHBauer commented:
Estimator.fit and Transformer.transform use a shared mechanism for copying parameters, both call Params.copy, the bulk of its implementation is in _copyValues.  This is the right place to do this.  I observe that fit and transform have already checked that the extra params are in a dict, so that check is superfluous.
This doesn't change the published API, since the requirement that values be param objects is not documented.

dongjoon-hyun commented on .github/PULL_REQUEST_[SPARK-29691]:
```diff
@@ -0,0 +1,37 @@
+
+### How was this patch tested?
+
+Added method test_copy_param_extras_check to test_param.py.
```
@JohnHBauer . Please remove this file . You already describe this information correctly.

srowen commented on python/pyspark/ml/param/__init__.py:
```diff
@@ -484,8 +484,19 @@ def _copyValues(self, to, extra=None):
+                    paramMap[param] = value
+                elif self.hasParam(param):
+                    # set value by name if vaiid
+                    paramMap[self.getParam(param)] = value
```
I don't know if we should support this, as it isn't specified this way (by name) elsewhere. Part of the reason is that it can be ambiguous if you have multiple instances of the same transformer in the pipeline? I'm only surprised it wouldn't already fail, so indeed maybe the validation below is reasonable.

JohnHBauer replied:
It can't be ambiguous in a pipeline: since `stages` is the only parameter that pipeline has, that is the only parameter that could be set this way.  If one did pass stages as extra parameters, it would have the same effect as creating a new pipeline with those stages.  No idea why someone would do that, but no compelling reason to stop them.

srowen replied:
Hm yeah I don't know if there's a use case for passing parameters to elements of a Pipeline this way. It might work, not sure. Anyway I think argument from consistency probably still stands.

BryanCutler commented:
My original suggestion was to only add an error message if the user did not pass in an instance of `Param` to the `extra` arg in `fit`. I don't think we should allow using strings because there would be a consistency problem with the rest of the param APIs, and like @srowen pointed out, there could be an ambiguity with pipelines. When calling `pipeline.fit` with `extra`, I believe you can pass in a set of params for all stages, and there could be duplicate stages or stages that have the same param name.

JohnHBauer committed:
Remove string values, check ParamGrid

Remove attempt to set values by name, added a check to ParamGridBuilder, rather than wait for fit to trigger an error

JohnHBauer committed:
Update tuning.py

Fix line length

JohnHBauer commented:
- Added a check to ParamGridBuilder.addGrid to enforce that keys be Param instances.  Without the check, it's a generic CartesianProductBuilder.  (The check in param.__init__._copyValues will also catch them, but only once fit is called.)
- Removed string usage.  (Validation of extra params would have had to be done by the first call to fit/transform.  Doing the check in _copyValues is more robust.)

BryanCutler commented on python/pyspark/ml/param/__init__.py:
```diff
@@ -484,8 +484,16 @@ def _copyValues(self, to, extra=None):
+                if isinstance(param, Param):
+                    paramMap[param] = value
+                else:
+                    raise ValueError("Expecting a valid instance of Param, but received: {}"
```
This should be a `TypeError` since it is due to type checking

BryanCutler commented on python/pyspark/ml/tuning.py:
```diff
@@ -88,8 +88,14 @@ def __init__(self):
+        if isinstance(param, Param):
+            self._param_grid[param] = values
+        else:
+            raise ValueError("param must be an instance of Param")
```
ditto, use `TypeError` here

BryanCutler commented on python/pyspark/ml/tuning.py:
```diff
@@ -88,8 +88,14 @@ def __init__(self):
+        (such as Estimator or Transformer).
         """
-        self._param_grid[param] = values
+        if isinstance(param, Param):
```
can you add a test for this too?

BryanCutler reviewed (commented):
Thanks @JohnHBauer , looks pretty good. I just requested a few minor changes.

JohnHBauer committed:
Added TypeError tests

JohnHBauer committed:
Update test_tuning.py

Linter found whitespace

BryanCutler commented:
ok to test

SparkQA commented:
**[Test build #114115 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114115/testReport)** for PR 26527 at commit [`b55254a`](https://github.com/apache/spark/commit/b55254ac7cb92d3641271c8a22010d7775c5e267).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18973/
Test PASSed.

SparkQA commented:
**[Test build #114115 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114115/testReport)** for PR 26527 at commit [`b55254a`](https://github.com/apache/spark/commit/b55254ac7cb92d3641271c8a22010d7775c5e267).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114115/
Test FAILed.

JohnHBauer committed:
Update test_param.py

SparkQA commented:
**[Test build #114117 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114117/testReport)** for PR 26527 at commit [`1081991`](https://github.com/apache/spark/commit/1081991242308f09f2e34f5cc7215424a845c2b4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18975/
Test PASSed.

SparkQA commented:
**[Test build #114117 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114117/testReport)** for PR 26527 at commit [`1081991`](https://github.com/apache/spark/commit/1081991242308f09f2e34f5cc7215424a845c2b4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114117/
Test PASSed.

BryanCutler reviewed (approved):
LGTM

BryanCutler commented:
merged to master, thanks @JohnHBauer !
